apps/mymedicare_cb/models.py

```diff
@@ -1,76 +1,120 @@
-                    for n in names:
-                        if n['use'] == 'usual':
-                            last_name = n['family']
-                            first_name = n['given'][0]
-                        if last_name or first_name:
-                            user.first_name = first_name
-                            user.last_name = last_name
-                            user.save()
-    except (UpstreamServerException, exceptions.NotFound):
-        logger.error("Failed to connect Beneficiary "
-                     "to FHIR")
+        if first_name == "" or last_name == "":
+            first_name, last_name = extract_beneficiary_names(backend_data)
+    except exceptions.NotFound:
+        fhir_id = None
 
+    fhir_source = get_resourcerouter()
+
+    user = create_beneficiary_record(username=subject,
+                                     user_id_hash=hicn_hash,
+                                     fhir_id=fhir_id,
+                                     fhir_source=fhir_source,
+                                     first_name=first_name,
+                                     last_name=last_name,
+                                     email=email)
     return user
 
 
+# TODO default empty strings to null, requires non-null constraints to be fixed
+def create_beneficiary_record(username=None,
+                              user_id_hash=None,
+                              fhir_id="",
+                              fhir_source=None,
+                              first_name="",
+                              last_name="",
+                              email=""):
+    assert username is not None
+    assert user_id_hash is not None
+
+    if User.objects.filter(username=username).exists():
+        raise ValidationError("user already exists", username)
+
+    if Crosswalk.objects.filter(user_id_hash=user_id_hash).exists():
+        raise ValidationError("user_id_hash already exists", user_id_hash)
+
+    if fhir_id and Crosswalk.objects.filter(fhir_id=fhir_id).exists():
+        raise ValidationError("fhir_id already exists", fhir_id)
+
+    with transaction.atomic():
+        user = User(username=username,
+                    first_name=first_name,
+                    last_name=last_name,
+                    email=email)
+        user.set_unusable_password()
+        user.save()
+        Crosswalk.objects.create(user=user,
+                                 fhir_source=fhir_source,
+                                 user_id_hash=user_id_hash,
+                                 fhir_id=fhir_id)
+
+        # Extra user information
+        # TODO: remvoe the idea of UserProfile
+        UserProfile.objects.create(user=user, user_type='BEN')
+        # TODO: magic strings are bad
+        group = Group.objects.get(name='BlueButton')  # TODO: these do not need a group
+        user.groups.add(group)
+    return user
+
+
+def extract_beneficiary_names(patient_resource):
+    if 'entry' in patient_resource:
```

Comment:
Thoughts on logging when data gets supplemented by the backend during user creation? I can imagine a case where we are trying to track down a user with unexpected or inconsistent fields and being thrown off by not knowing where that data came from. Maybe this would be evident from logs we already have though.

Reply:
removed for now, I've added logs around what sls adds to the bb20 system